### PR TITLE
Refactored RepoPath to be a proper MutableSequence

### LINKED ISSUE #4236: Spack complains of missing patch file when staging hsf sqlite package
This appears to be an incomplete override of builtin.sqlite by hsf.sqlite. The hsf.sqlite package does not have a patch. I appears to pick up newer version from hsf.sqlite which does have a patch.

 spack stage hsf.sqlite
==> Error: No such patch file for package sqlite: /home/gartung/hep-spack/packages/sqlite/sqlite_b0.patch

If the hsf repo is removed then 
spack stage sqlite 
works as expected.

### LINKED ISSUE #5002: Patches and Package Namespace
I have another repo other than builtin.  There is a lulesh package in both my repo and the builtin repo.  The builtin spackage has a patch, the version in my repo does not.  Spack was looking for the patch in my repo and failing to find it.  My thought was that the way spack checks for patches ignores the namespace?  
However after installing the patch into my repo directory to work then removing it to recreate the error solved the issue.

### PR DESCRIPTION
**Pull request by alalazo**
I spent the last couple of days studying the code that implements repositories (and I have to say it has been quite instructive to see how it blends into the python "import" protocol). While I was at it I thought it could have been a good exercise to refactor some parts to be more "pythonic"  hence this PR.

The modifications here involve just `RepoPath`, as I would like to keep the PR as focused as possible. If there's agreement on the type of changes I will start having a look at `Repo` and `NamespaceTrie` in other PRs (in case they need any change).

Let me know if you agree with these kind of PRs or if you can figure out further tests besides the ones implemented.

Fixes #4236 
Fixes #5002

### CODE REVIEW

**Comment by adamjstewart** [comment]
While you're digging through our repo logic, you might want to take a look at #4236 and #1801.

**Comment by alalazo** [comment]
@tgamblin @scheibelp ping

**Comment by alalazo** [comment]
@gartung This should solve your problem in #4236 Can you confirm?

**Comment by gartung** [comment]
The patch fixes the problem I was seeing.

**Comment by alalazo** [comment]
@scheibelp @tgamblin If you agree with the changes here it's +0.13% coverage and #4236 fixed...

**Comment by alalazo** [comment]
ping 

(In case you don't agree with the PR, I think the last commit here can be cherry-picked seamlessly to solve #4236)

**Comment by alalazo** [comment]
ping

**Comment by alalazo** [comment]
@scheibelp Do you have time to take a look at this PR?

It shouldn't break any internal, as it is a small refactoring of `RepoPath` to be more pythonic. I think overall it cleans the code that was using `spack.repo`, especially in the tests where it's not needed anymore to "swap". 

The only choice which could maybe be controversial is the fact that I created a module `defaults.py` which mixes configuration files and object creation (instead of having that logic in the class `__init__`).

**Comment by scheibelp** [comment]
I'll review this by the end of 8/8

**Comment by alalazo** [comment]
@scheibelp Thanks!

File: lib/spack/spack/test/repository.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+class TestRepoPath(object):
+
+    def test_error_during_construction(self):
+
+        # If the repository has no repo.yaml, raise BadRepoError
+        with pytest.raises(spack.repository.BadRepoError):
+            spack.repository.RepoPath('/foo/bar')
+
+    def test_magic_methods(self, mock_repository):
```
Comment (scheibelp) [inline lib/spack/spack/test/repository.py]: More-descriptive name is preferred. This also appears to be testing several different behaviors; for example deleting, copying, and adding could be tested separately.

**Comment by scheibelp** [comment]
Oh and I should add: while I added comments on updating RepoPath to be a MutableSequence I was hoping to discuss it more before you did additional work on that part (for example updating the unit testing)  since I don't understand the benefits at the moment.

File: lib/spack/spack/patch.py [hunk 2]
```diff
@@ -90,7 +91,8 @@ class FilePatch(Patch):
     def __init__(self, pkg, path_or_url, level):
         super(FilePatch, self).__init__(pkg, path_or_url, level)
 
-        pkg_dir = spack.repo.dirname_for_package_name(pkg.name)
+        m = inspect.getmodule(pkg)
```
Comment (alalazo) [inline lib/spack/spack/patch.py]: > The commit cf65636 appears to contain all changes which account for fixing #4236

Yes, there was even a comment above:

> (In case you don't agree with the PR, I think the last commit here can be cherry-picked seamlessly to solve #4236)

So: should I move this to another PR and fix `spack.repo.dirname_for_package` to take a `Package` instead of a string?

Comment (scheibelp) [inline lib/spack/spack/patch.py]: > Yes, there was even a comment above:

Yep. I missed that initially. I prefer the initial summary to be updated rather than applying all deltas in the comments (although I definitely should have been more careful).

> So: should I move this to another PR and fix spack.repo.dirname_for_package to take a Package instead of a string?

That is my preference

**Comment by alalazo** [comment]
> Oh and I should add: while I added comments on updating RepoPath to be a MutableSequence I was hoping to discuss it more before you did additional work on that part (for example updating the unit testing) since I don't understand the benefits at the moment.

By its own admission:
```python
"""A RepoPath *is a list of repos* that function as one.
...
"""
```
Using a `MutableSequence` defines a stronger contract between the class and its clients, and supports implementing it (in the sense that you get a lot for free, if you implement the methods that are mandated by the `abc`). If you take a look at what we have now in `develop`:
```python
if args.mock:
    from spack.repository import RepoPath
    spack.repo.swap(RepoPath(spack.mock_packages_path))
...

v = spack.repo.first_repo()
...
spack.repo.put_last(x)
```
I think you may agree that this *is not really a list* - but kind of. After the PR the three above look like:
```python
if args.mock:
    spack.repo.clear()
    spack.repo.append_from_path(spack.mock_packages_path)
...
v = spack.repo[0]
...
spack.repo.append(x)
```

On top of that you'll get for free common idioms for lists:
```python
l = len(spack.repo)
...
for x in spack.repo:
   # do something
...
s = reverse(sorted(spack.repo))
```
etc. So to your question "Why?" I reply "Why not?". This is certainly better than implementing only part of what 'makes' a list, and try essentially to bake our incomplete `MutableSequence`. In the end I believe that adhering to the `MutableSequence` class will result in a more familiar interface and will give more flexibility in the way we can manipulate objects.

File: lib/spack/spack/repository.py [hunk 3]
```diff
@@ -112,48 +128,138 @@ def __init__(self, *repo_dirs, **kwargs):
...
+    def append_from_path(self, root):
+        """Append a repository to the list of managed ones. Takes a path
+        as argument.
+
+        Args:
+            root: path to the repository
+        """
+        try:
+            repo = Repo(root, self.super_namespace)
+            self.append(repo)
```
Comment (alalazo) [inline lib/spack/spack/repository.py]: With an IDE this is a <kbd>Ctrl</kbd>+<kbd>B</kbd> away:
```python
def append(self, value):
    'S.append(value) -- append value to the end of the sequence'
    self.insert(len(self), value)
```
> However while it shouldn't make this class harder to use, IMO it does make it harder to modify.

It does make the class *easier to use*, as it is now adhering to a stronger contract - that of `MutableSequence`s. I don't get you point when you say it will be harder to modify. Why?

File: lib/spack/spack/patch.py (hunk 2 above)
Comment (alalazo) [inline lib/spack/spack/patch.py]: @scheibelp Ok. Will do this tomorrow.

File: lib/spack/spack/defaults.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+"""Creates various objects that are part of Spack Core according to the
```
Comment (alalazo) [inline lib/spack/spack/defaults.py]: I think you could see the thing also from the other side: having a module like this one somewhere is a starting point to refactor other modules. But that's definitely fine with me to keep the function in `spack.repo` if you think it's better (ot needs to be discussed further).

File: lib/spack/spack/test/directory_layout.py [hunk 5]
```diff
@@ -196,7 +196,10 @@ def test_handle_unknown_package(
         layout.create_install_directory(spec)
         installed_specs[spec] = layout.path_for_spec(spec)
 
-    spack.repo.swap(mock_db)
+    # Plug-in the mock repository
+    cache_real_repo = spack.repo[:]
```
Comment (alalazo) [inline lib/spack/spack/test/directory_layout.py]: You shouldn't count only methods I implemented directly. :smile:  

I use `clear` (whose implementation is ported from python 3.6) and `append` instead of an in-place swap for instance. I use for loops over the repo. The point is: once you blend into the usual list syntax, you'll start using it.

File: lib/spack/spack/repository.py (hunk 3 above)
Comment (scheibelp) [inline lib/spack/spack/repository.py]: I should preface by saying I'm not sure why ```RepoPath.swap``` ever had to exist - why couldn't ```conftest.builtin_mock``` have said something like

```
saved_repo = spack.repo
spack.repo = mock_db
```

If something more complex is necessary, and if the MutableSequence functionality makes this easier, then I would like the change more.

In my opinion inheriting from MutableSequence makes it harder to modify because it's harder to understand how it works. It is true that an IDE may retrieve the appropriate function, but you still have to mentally combine this sequence of function calls into the desired end result.

> It does make the class easier to use, as it is now adhering to a stronger contract

Speaking generically, enforcing the stronger contract is part of what makes it harder to modify: when/if someone edits RepoPath they must consider whether their operations satisfy the requirements placed on a MutableSequence.

It can be worthwhile to increase complexity if other things are made simpler, but the only simplification I saw was eliminating the call to ```.swap```. In my opinion there ought to be a way to achieve that simplification that would avoid introducing the complexity of satisfying an additional contract.

Comment (alalazo) [inline lib/spack/spack/repository.py]: > In my opinion inheriting from MutableSequence makes it harder to modify because it's harder to understand how it works.

and

> Speaking generically, enforcing the stronger contract is part of what makes it harder to modify

Speaking generically, that may be true (and that's probably why you are rarely advised to come up with your own abc).

In this case, instead, enforcing a contract that makes a class *behave like a built-in structure of the language* results in code that is easier to modify, because you know what to expect. Besides `swap`, in the current implementation we have:
```python
RepoPath.put_first(x) # Is it like [].insert(0, x) ?
RepoPath.put_last(x)  # Is it like [].append(x) ?
RepoPath.remove(x) # Is it like [].remove(x) ?
RepoPath.first_repo() # Is it the same as [].__getitem__(0) ?
```
With a class that resembles a list sometimes, but does not other times, following your reasoning we'll need to inspect the semantic of every function before modifying one. With a `MutableClass` we don't: as soon as we implement the prescribed abstract functions with the same semantic as list, every other method that is part of the protocol is already implemented and works "as expected", i.e. as a list. 

A counter-example on the fact that it's not so difficult to modify this class (as compared to the previous implementation) is that the different semantic for `__contains__` has been ported just maintaining the override, and making the new, unexpected behavior explicit in the docs.

Another example focused on encapsulation: with the current interface I can construct a list of repos, and then get back only the first element through its interface. To get other elements or iterate over them I need to access "implementation details" of my object, like `spack.repo.repos`. That wouldn't be bad per-se in python, but in this case we have three states that must be kept consistent (`repos`, `by_path`, `by_namespace`), and we are placing this responsibility on clients of `RepoPath`. Using a `MutableSequence` we can operate on the object API directly and modify its state in a consistent way, like if it was a list.

In the end for me this is all about:

- having a consistent interface across the code, for behaviors that are similar (list, dictionaries, tuples, etc.)
- improving encapsulation of the data structure we use

You're right that these behaviors are not yet used in the code, but probably we'll need to touch repos to solve #3115. This modification in my opinion is a step toward making `RepoPath` more flexible to work with, and it doesn't hurt (at least imo).

File: lib/spack/spack/defaults.py (hunk 4 above)
Comment (alalazo) [inline lib/spack/spack/defaults.py]: On a more practical side: in case the `MutableSequence` part needs to be discussed further, or we decide for any reason not to use it, let me know if you want me to extract / recode this modification as a separate PR.

File: lib/spack/spack/repository.py (hunk 3 above)
Comment (scheibelp) [inline lib/spack/spack/repository.py]: > In this case, instead, enforcing a contract that makes a class behave like a built-in structure of the language results in code that is easier to modify, because you know what to expect.

I disagree there. As a user of lists I know what to expect. As someone mucking around inside the internals of the implementation of ```MutableSequence``` I don't know what to expect. Will I accidentally override a method? Will I forget to call a ```_private_method```? Are there data structures inside the implementation of ```MutableSequence``` which need to be maintained? If I want this to behave correctly as a ```MutableSequence``` then I have to be familiar with its implementation.

> as soon as we implement the prescribed abstract functions with the same semantic as list, every other method that is part of the protocol is already implemented and works "as expected", i.e. as a list.

The most important function from a user perspective for ```RepoPath``` is ```get```. Beyond that, adding additional behaviors to a class that doesn't need them may be confusing. Will a user want to do ```repo_path[4:9]```? What does it mean in terms of the outcome of ```RepoPath.get``` if I say ```repo_path.insert(3, x)```?

Also, knowing that ```RepoPath``` is a mutable sequence won't tell a user for example whether ```repo_path[0]``` or ```repo_path[-1]``` is the most important repo: MutableSequence doesn't have anything to say about order and although a user could guess, they are advised to understand the interface. ```RepoPath.first_repo``` is more suggestive of priority.

> With a class that resembles a list sometimes, but does not other times, following your reasoning we'll need to inspect the semantic of every function before modifying one.

I think this is generally true: when you add functionality to a class you need to understand its internal implementation. The exception is when you are using composition. That requirement still exists when inheriting from ```MutableSequence```

> probably we'll need to touch repos to solve #3115

I read through that and didn't get why repos were the issue there. Assuming they are, I'd prefer these sorts of changes be integrated into a PR directed at fixing that in order to demonstrate the usefulness of those changes.

Comment (alalazo) [inline lib/spack/spack/repository.py]: Ok, got the message: *goose-typing* not welcome here. :smile: I'll close this PR as soon as I extract the fix. Anyhow:

> If I want this to behave correctly as a MutableSequence then I have to be familiar with its implementation.

Not True. You have to be familiar with its contract, i.e. you need to know the semantic of the abstract functions and the other parts will follow. Or, said otherwise, this is as true as saying that you need to know the details of Tim sort every time you pass a `key` to `sorted`. I personally don't.
